File: smallpict-php/src/Helper/SmallPictHelper.php
<?php

namespace SmallPict\Helper;

use SmallPict\Client;
use SmallPict\Config as SdkConfig;
use SmallPict\Models\OptimizeOptions;
use SmallPict\Config\Config;
use SmallPict\Database\Database;
use GuzzleHttp\Client as GuzzleClient;

class SmallPictHelper
{
    public static function processUpload(?array $fileItem, ?string $base64Input, ?string $customFilename): array
    {
        $data = null;
        $filename = $customFilename ?? '';
        $mimeType = 'image/jpeg';

        if (!empty($base64Input)) {
            $b64 = trim($base64Input);
            if (str_contains($b64, ';base64,')) {
                $parts = explode(';base64,', $b64, 2);
                if (str_starts_with($parts[0], 'data:')) {
                    $mimeType = substr($parts[0], 5);
                }
                $b64 = $parts[1];
            } elseif (str_starts_with($b64, 'data:') && str_contains($b64, ',')) {
                $parts = explode(',', $b64, 2);
                $mimeType = explode(';', substr($parts[0], 5))[0];
                $b64 = $parts[1];
            }

            $b64 = preg_replace('/\s+/', '', $b64);
            $decoded = base64_decode($b64, true);
            if ($decoded === false) {
                throw new \InvalidArgumentException('Invalid base64 image data');
            }
            $data = $decoded;
            if (empty($filename)) {
                $filename = 'image' . self::mimeToExtension($mimeType);
            }
        } elseif (!empty($fileItem) && isset($fileItem['tmp_name']) && file_exists($fileItem['tmp_name'])) {
            $data = file_get_contents($fileItem['tmp_name']);
            if (empty($filename)) {
                $filename = $fileItem['name'] ?? 'image.jpg';
            }
            if (!empty($fileItem['type'])) {
                $mimeType = $fileItem['type'];
            }
        } else {
            throw new \InvalidArgumentException('No image content provided in request');
        }

        if (empty($data)) {
            throw new \InvalidArgumentException('Image data is empty');
        }

        $baseName = pathinfo($filename, PATHINFO_FILENAME);
        $ext = pathinfo($filename, PATHINFO_EXTENSION);
        $ext = $ext ? '.' . $ext : self::mimeToExtension($mimeType);
        if (empty($baseName)) {
            $baseName = 'image';
        }

        $epochSec = time();
        $timestampedFilename = "{$baseName}_{$epochSec}{$ext}";
        $imageId = sprintf(
            '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
            mt_rand(0, 0xffff), mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0x0fff) | 0x4000,
            mt_rand(0, 0x3fff) | 0x8000,
            mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
        );

        $sizeOrigin = strlen($data);
        $size = $sizeOrigin;
        $imageUrl = '';

        $config = Config::getInstance();
        $mode = strtolower($config->smallPictMode);
        $hasCredentials = !empty($config->smallPictApiKey) && !empty($config->smallPictSecretKey);

        if ($hasCredentials) {
            $sdkConfig = new SdkConfig($config->smallPictApiKey, $config->smallPictSecretKey, $config->smallPictBaseUrl);
            $client = new Client($sdkConfig);

            $optOptions = new OptimizeOptions(
                $config->smallPictFormat,
                $config->smallPictQuality,
                null,
                null,
                'cover',
                false,
                true,
                $timestampedFilename,
                $mimeType
            );

            $stream = fopen('php://temp', 'r+');
            fwrite($stream, $data);
            rewind($stream);

            $result = $client->optimize($stream, $optOptions);
            if (is_resource($stream)) {
                fclose($stream);
            }

            if ($result->getUploadUrl()) {
                $ch = curl_init($result->getUploadUrl());
                curl_setopt_array($ch, [
                    CURLOPT_CUSTOMREQUEST  => 'PUT',
                    CURLOPT_POSTFIELDS     => $data,
                    CURLOPT_HTTPHEADER     => ["Content-Type: {$mimeType}"],
                    CURLOPT_RETURNTRANSFER => true,
                    CURLOPT_TIMEOUT        => 60,
                    CURLOPT_HTTP_VERSION   => CURL_HTTP_VERSION_1_1,
                    CURLOPT_IPRESOLVE      => CURL_IPRESOLVE_V4,
                ]);
                $putRes = curl_exec($ch);
                $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
                $err = curl_error($ch);
                curl_close($ch);

                if ($httpCode < 200 || $httpCode >= 300 || $putRes === false) {
                    throw new \RuntimeException("SmallPict S3 Upload failed HTTP {$httpCode}: {$err}");
                }
            }

            $remoteUrl = $result->getUrl();
            $jobId = $result->getJobId();
            $status = strtolower($result->getStatus());

            if (!empty($jobId) && (in_array($status, ['pending', 'processing', 'queued']) || empty($remoteUrl))) {
                for ($i = 0; $i < 20; $i++) {
                    usleep(2000000); // 2 seconds
                    $jobStatus = $client->getJobStatus($jobId);
                    $st = strtolower($jobStatus->getStatus());
                    if (in_array($st, ['completed', 'succeeded', 'success', 'ready', 'done'])) {
                        if ($jobStatus->getUrl()) {
                            $remoteUrl = $jobStatus->getUrl();
                        }
                        if ($jobStatus->getBytesSaved() && $jobStatus->getBytesSaved() > 0) {
                            $size = $sizeOrigin - $jobStatus->getBytesSaved();
                        }
                        break;
                    }
                }
            }

            if (empty($remoteUrl)) {
                throw new \RuntimeException('Failed to get optimized image URL from SmallPict');
            }

            if ($mode === 'locale' || $mode === 'local') {
                $cleanUrlPath = str_contains($remoteUrl, '?') ? substr($remoteUrl, 0, strpos($remoteUrl, '?')) : $remoteUrl;
                $remoteExt = strtolower(pathinfo($cleanUrlPath, PATHINFO_EXTENSION));
                if ($remoteExt === 'webp') {
                    $timestampedFilename = "{$baseName}_{$epochSec}.webp";
                    $mimeType = 'image/webp';
                } elseif ($remoteExt === 'avif') {
                    $timestampedFilename = "{$baseName}_{$epochSec}.avif";
                    $mimeType = 'image/avif';
                } elseif ($remoteExt === 'png') {
                    $timestampedFilename = "{$baseName}_{$epochSec}.png";
                    $mimeType = 'image/png';
                } elseif (in_array($remoteExt, ['jpg', 'jpeg'])) {
                    $timestampedFilename = "{$baseName}_{$epochSec}.jpg";
                    $mimeType = 'image/jpeg';
                }

                $uploadDir = $config->uploadPath;
                if (!is_dir($uploadDir)) {
                    mkdir($uploadDir, 0755, true);
                }

                $imageData = self::downloadImage($remoteUrl);
                $destPath = rtrim($uploadDir, '/') . '/' . $timestampedFilename;
                file_put_contents($destPath, $imageData);

                $imageUrl = "/uploads/{$timestampedFilename}";
                $size = filesize($destPath);
            } else {
                $imageUrl = $remoteUrl;
            }
        } else {
            $uploadDir = $config->uploadPath;
            if (!is_dir($uploadDir)) {
                mkdir($uploadDir, 0755, true);
            }

            $destPath = rtrim($uploadDir, '/') . '/' . $timestampedFilename;
            file_put_contents($destPath, $data);
            $imageUrl = "/uploads/{$timestampedFilename}";
        }

        $record = [
            'id'          => $imageId,
            'filename'    => $timestampedFilename,
            'url'         => $imageUrl,
            'size'        => $size,
            'size_origin' => $sizeOrigin,
            'mime_type'   => $mimeType,
            'created_at'  => date('Y-m-d H:i:s'),
        ];

        $pdo = Database::getConnection();
        if ($pdo !== null) {
            $stmt = $pdo->prepare("
                INSERT INTO public.images (id, filename, url, size, size_origin, mime_type, created_at)
                VALUES (:id, :filename, :url, :size, :size_origin, :mime_type, :created_at)
            ");
            $stmt->execute($record);
        }

        return $record;
    }

    private static function downloadImage(string $url): string
    {
        $lastError = '';
        for ($i = 0; $i < 3; $i++) {
            if ($i > 0) {
                usleep(500000); // 0.5s retry delay
            }

            $ch = curl_init($url);
            curl_setopt_array($ch, [
                CURLOPT_RETURNTRANSFER => true,
                CURLOPT_FOLLOWLOCATION => true,
                CURLOPT_TIMEOUT        => 60,
                CURLOPT_HTTP_VERSION   => CURL_HTTP_VERSION_1_1,
                CURLOPT_IPRESOLVE      => CURL_IPRESOLVE_V4,
                CURLOPT_SSL_VERIFYPEER => true,
                CURLOPT_SSL_VERIFYHOST => 2,
            ]);

            $response = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            $curlError = curl_error($ch);
            curl_close($ch);

            if ($response !== false && $httpCode >= 200 && $httpCode < 300) {
                return $response;
            }

            $lastError = $curlError ?: "HTTP {$httpCode}";
        }

        // Fallback using stream context
        $context = stream_context_create([
            'http' => [
                'timeout'          => 60,
                'protocol_version' => 1.1,
                'ignore_errors'    => true,
            ],
            'ssl' => [
                'verify_peer'      => false,
                'verify_peer_name' => false,
            ],
        ]);
        $streamContent = @file_get_contents($url, false, $context);
        if ($streamContent !== false && !empty($streamContent)) {
            return $streamContent;
        }

        throw new \RuntimeException("Failed to download compressed image from SmallPict: {$lastError}");
    }

    private static function mimeToExtension(string $mimeType): string
    {
        return match (strtolower($mimeType)) {
            'image/png' => '.png',
            'image/webp' => '.webp',
            'image/avif' => '.avif',
            'image/gif' => '.gif',
            'image/svg+xml' => '.svg',
            default => '.jpg',
        };
    }
}

File: smallpict-ruby/app/helpers/smallpict_helper.rb
require 'base64'
require 'securerandom'
require 'fileutils'
require 'httparty'
require 'json'
require 'stringio'
require 'smallpict'
require_relative '../config'
require_relative '../models/image'

module SmallPictHelper
  def self.mime_to_ext(mime)
    case mime.to_s.downcase
    when 'image/png' then '.png'
    when 'image/webp' then '.webp'
    when 'image/gif' then '.gif'
    when 'image/avif' then '.avif'
    else '.jpg'
    end
  end

  def self.extract_data(file_hash: nil, base64_str: nil, filename: nil)
    data = nil
    fname = filename.to_s.delete("\0").strip
    mime = 'image/jpeg'

    if base64_str && !base64_str.strip.empty?
      b64 = base64_str.strip
      if b64.include?(';base64,')
        parts = b64.split(';base64,', 2)
        mime = parts[0].sub(/^data:/, '') if parts[0].start_with?('data:')
        b64 = parts[1]
      elsif b64.start_with?('data:') && b64.include?(',')
        parts = b64.split(',', 2)
        mime = parts[0].sub(/^data:/, '').split(';')[0]
        b64 = parts[1]
      end

      clean_b64 = b64.gsub(/\s+/, '')
      data = Base64.decode64(clean_b64)
      fname = "image#{mime_to_ext(mime)}" if fname.empty?
    elsif file_hash && file_hash[:tempfile]
      data = file_hash[:tempfile].read
      fname = file_hash[:filename].to_s.delete("\0").strip if fname.empty?
      mime = file_hash[:type] if file_hash[:type]
    end

    raise 'Image data is empty' if data.nil? || data.empty?

    [data, fname, mime]
  end

  def self.upload(file_hash: nil, base64_str: nil, original_filename: nil)
    data, orig_fname, mime_type = extract_data(
      file_hash: file_hash,
      base64_str: base64_str,
      filename: original_filename
    )

    clean_orig_fname = orig_fname.to_s.delete("\0").strip
    base = File.basename(clean_orig_fname, '.*')
    base = 'image' if base.empty?
    ext = File.extname(clean_orig_fname)
    ext = mime_to_ext(mime_type) if ext.empty?

    epoch_sec = Time.now.to_i
    timestamped_filename = "#{base}_#{epoch_sec}#{ext}"
    image_id = SecureRandom.uuid
    size_origin = data.bytesize
    size = size_origin
    image_url = ''
    mode = AppConfig.smallpict_mode

    has_credentials = !AppConfig.smallpict_api_key.empty? && !AppConfig.smallpict_secret_key.empty?

    if has_credentials
      client = SmallPict::Client.new(
        api_key: AppConfig.smallpict_api_key,
        secret_key: AppConfig.smallpict_secret_key,
        base_url: AppConfig.smallpict_base_url
      )

      result = client.optimize(
        StringIO.new(data),
        filename: timestamped_filename,
        mime_type: mime_type,
        format: AppConfig.smallpict_format,
        quality: AppConfig.smallpict_quality
      )

      if result.upload_url && !result.upload_url.empty?
        put_res = HTTParty.put(
          result.upload_url,
          body: data,
          headers: { 'Content-Type' => mime_type },
          timeout: 60
        )
        raise "SmallPict S3 upload failed with HTTP #{put_res.code}" if put_res.code >= 300
      end

      remote_url = result.url.to_s
      job_id = result.job_id
      status = result.status.to_s.downcase

      if job_id && (%w[pending processing queued].include?(status) || remote_url.empty?)
        20.times do
          sleep 2
          begin
            job_data = client.get_job_status(job_id)
            job_status = job_data.status.to_s.downcase
            if %w[succeeded completed success ready done].include?(job_status)
              remote_url = job_data.url.to_s if job_data.url
              if job_data.bytes_saved && job_data.bytes_saved.to_i > 0
                size = size_origin - job_data.bytes_saved.to_i
              end
              break
            elsif %w[failed error].include?(job_status)
              raise "SmallPict optimization failed: #{job_data.error || 'unknown error'}"
            end
          rescue StandardError => e
            raise e if e.message.start_with?('SmallPict optimization failed')
          end
        end
      end

      raise 'Failed to get optimized image URL from SmallPict' if remote_url.empty?

      if %w[locale local].include?(mode)
        clean_url = remote_url.include?('?') ? remote_url.split('?').first : remote_url
        remote_ext = File.extname(clean_url).downcase
        if remote_ext == '.webp'
          timestamped_filename = "#{base}_#{epoch_sec}.webp"
          mime_type = 'image/webp'
        elsif remote_ext == '.avif'
          timestamped_filename = "#{base}_#{epoch_sec}.avif"
          mime_type = 'image/avif'
        elsif remote_ext == '.png'
          timestamped_filename = "#{base}_#{epoch_sec}.png"
          mime_type = 'image/png'
        elsif %w[.jpg .jpeg].include?(remote_ext)
          timestamped_filename = "#{base}_#{epoch_sec}.jpg"
          mime_type = 'image/jpeg'
        end

        FileUtils.mkdir_p(AppConfig.upload_path)
        dl_res = HTTParty.get(remote_url, timeout: 60)
        raise "Failed to download compressed image HTTP #{dl_res.code}" if dl_res.code >= 300

        dest_path = File.join(AppConfig.upload_path, timestamped_filename)
        File.binwrite(dest_path, dl_res.body)

        image_url = "/uploads/#{timestamped_filename}"
        size = dl_res.body.bytesize
      else
        image_url = remote_url
      end
    else
      FileUtils.mkdir_p(AppConfig.upload_path)
      dest_path = File.join(AppConfig.upload_path, timestamped_filename)
      File.binwrite(dest_path, data)
      image_url = "/uploads/#{timestamped_filename}"
    end

    Image.create(
      id: image_id,
      filename: timestamped_filename,
      url: image_url,
      size: size,
      size_origin: size_origin,
      mime_type: mime_type,
      created_at: Time.now
    )
  end
end
